lib/doorkeeper/oauth/helpers/scope_checker.rb

```diff
@@ -18,26 +18,35 @@ def valid?
           end
 
           def match?
-            valid? && parsed_scopes.has_scopes?(@valid_scopes)
+            valid? && parsed_scopes.equal?(@valid_scopes)
           end
 
           private
 
-          def valid_scopes(server_scopes, application_scopes)
-            if application_scopes.present?
-              server_scopes & application_scopes
+          def valid_scopes(scopes_source, application)
+            # TODO: Refactor this
```

Comment:
Let's address now this `TODO`. A first idea that come to mind is `return` early in positive cases, and leave as implicit return (outside of every conditional, at the end of the function) the `scopes_source.scopes` part.